RfidShelf/ShelfRfid.cpp

```diff
@@ -78,68 +87,130 @@ void ShelfRfid::handleRfid() {
-      char readFolder[18];
-      readFolder[0] = '/';
-      // allthough sizeof(readBuffer) == 18, we only get 16 byte of data (2 bytes are CRC)
-      memcpy(readFolder + 1 , readBuffer, 16 * sizeof(uint8_t) );
-      // readBuffer will already contain \0 if the folder name is < 16 chars, but otherwise we need to add it
-      readFolder[17] = '\0';
-
-      char currentFolder[101];
-      currentFolder[0] = '/';
-      _playback.currentFolder(currentFolder+1, sizeof(currentFolder)-1);
-      if ((_playback.playbackState() != PLAYBACK_NO) && (strcmp(readFolder, currentFolder) == 0)) {
-        Sprint("Resuming "); Sprintln(currentFolder);
-        _playback.resumePlayback();
-        _playback.playingByCard = true;
-      } else if (_playback.switchFolder(readFolder)) {
-        uint8_t configBuffer[18];
-        if (readRfidBlock(2, 0, configBuffer, sizeof(configBuffer)) && (configBuffer[0] == 137)) {
-          if(configBuffer[1] > 0) {
-            Sprint(F("Setting volume: ")); Sprintln(configBuffer[2]);
-            _playback.volume(configBuffer[2]);
-          }
-        } else {
-          _playback.volume(DEFAULT_VOLUME);
-          // "Upgrade" card
-          writeConfigBlock();
-        }
-
-        _playback.startPlayback();
-        _playback.playingByCard = true;
+  if (!readRfidBlock(1, 0, readBuffer, sizeof(readBuffer))) {
+    // we could not read card, skip handling
+    return;
+  }
+
+  // Store uid
+  memcpy(_lastCardUid, _mfrc522.uid.uidByte, 4 * sizeof(uint8_t) );
+
+  if(readBuffer[0] == '\0') {
+    if(_playback.switchFolder("/")) {
+      _playback.startFilePlayback("", "unknown_card.mp3");
+    }
+    return;
+  }
+
+  char readFolder[18];
+  readFolder[0] = '/';
+  // allthough sizeof(readBuffer) == 18, we only get 16 byte of data (2 bytes are CRC)
+  memcpy(readFolder + 1 , readBuffer, 16 * sizeof(uint8_t) );
+  // readBuffer will already contain \0 if the folder name is < 16 chars, but otherwise we need to add it
+  readFolder[17] = '\0';
+
+  strcpy(myCard.folder, readFolder);
+
+  char currentFolder[101];
+  currentFolder[0] = '/';
+  _playback.currentFolder(currentFolder+1, sizeof(currentFolder)-1);
+  if ((_playback.playbackState() != PLAYBACK_NO) && (strcmp(myCard.folder, currentFolder) == 0)) {
+    Sprint("Resuming "); Sprintln(currentFolder);
+    _playback.resumePlayback();
+    _playback.playingByCard = true;
+  } else if (_playback.switchFolder(myCard.folder)) {
+    handleRfidConfig();
+    _playback.setPlaybackOption(myCard.repeat, myCard.shuffle);
+    _playback.startPlayback();
+    _playback.playingByCard = true;
+  }
+
+  dumpCurrentCard();
+}
+
+void ShelfRfid::handleRfidConfig() {
+  uint8_t configBuffer[18];
+  if (!readRfidBlock(2, 0, configBuffer, sizeof(configBuffer))) {
+    // we could not read card config, skip handling
+    return;
+  }
+
+  myCard.repeat = true;
+  myCard.shuffle = false;
+  myCard.stopOnRemove = true;
+
+  if(configBuffer[0] < RFID_CONFIG_VERSION) {
```

Comment:
Ok. I thought that could be a nice step to introduce versioning of configuration options. 137 for volume-only config, 138 for volume and 3 new options, 139 for something next...
Shall I remove this?

Reply:
just make it a == and it should work fine. Versioning migh be necessary at some point but it might make sense to add a separate version field, in addition to the magic marker. And currently it's not needed at all, so let's leave the problem to future us :-D